whitepapers/0011_accountant.md

```diff
@@ -0,0 +1,190 @@
+- Error - This indicates that the accountant ran into an error while processing an observation.  The guardian should halt further processing of the transfer.  See the [Handling Rejections](https://www.notion.so/tbjump-copy-Accounting-v3-11572f647daf46f1a1e7d3ac80050ac0) section for more details
+
+#### Handling stuck transfers
+
+It may be possible that some transfers don’t receive enough observations to reach quorum.  This can happen if a number of guardians miss the transaction on the origin chain.  Since the accountant runs before the p2p gossip layer, it cannot take advantage of the existing automatic retry mechanisms that the guardian network uses.  To deal with this the accountant contract provides a `MissingObservations` query that returns a list of pending transfers for which the accountant does not have an observation from a particular guardian.
+
+Guardians are expected to periodically run this query and re-observe txs for which the accountant is missing observations from them.
+
+#### Contract Upgrades
+
+Upgrading the contract will be done via a governance action from the guardian network with the same mechanism we use for the existing wormhole cosmwasm contracts.
+
+### Account Management
+
+The accountant will keep track of the various token balances on each L1 and ensure that they never become negative.  When committing a transfer, the accountant will parse the tokenbridge payload to figure out the emitter chain, recipient chain, token chain, and token address.  These values will then be used to look up the relevant accounts and adjust the balances as needed.
+
+We will have the following accounts:
+
+- Accounts for native tokens on L1 chains ($SOL on solana, $ETH or $wETH on ethereum).  These are classified as assets because they represent actual tokens held by the smart contracts on the various chains.  All tokens that are not issued by the token bridge itself are considered native tokens, even if they are wrapped tokens issued by another bridge.  If the token is held by the bridge contract rather than burned, then it is an asset.  We’ll call these “custody accounts”.
+- Accounts for wrapped tokens (wrapped SOL on ethereum, wrapped ETH on solana).  These are classified as liabilities.  These tokens are not held by the smart contract and instead represent a liability for the bridge: if a user returns a wrapped token to the bridge, the bridge **must** be able to transfer the native token to that user on its native chain.  We’ll call these “wrapped accounts”.
+
+Each account is identified via a `(chain_id, token_chain, token_address)` tuple.  Custody accounts are those where `chain_id == token_chain`, while the rest are wrapped accounts.  Locking a native token or minting a wrapped token will increase the balance of the relevant account while unlocking a native token or burning a wrapped token will decrease the balance.  Transfers that would cause the balance of an account to become negative will be rejected.
```

Comment:
This explanation could use some rewording to make things more clear.

Reply:
best to refer to the section below

Reply:
More context added in bc888cdd92e403915e8c62da53563a93ba152315